File: javascript/beer-song/beer-song.js
var BeerSong = function() {}

BeerSong.prototype.verse = function(num) {
    return single_verse(num);
}

BeerSong.prototype.sing = function(start, end) {
    if (end === undefined) { end = 0; }
    let builder = "";
    for (let i = start; i >= end; i--) {
        builder += single_verse(i) + "\n";
    }
    return builder.substring(0, builder.length - 1);
}

var single_verse = function(num) {
    if (num == 0) { return last_verse; }
    let start_num = num;
    let start_bottle = num == 1 ? "bottle" : "bottles";
    let end_num = num - 1 == 0 ? "no more" : num - 1;
    let end_bottle = num - 1 == 1 ? "bottle" : "bottles";

    let first_constant = "of beer on the wall,";
    let second_constant = "of beer.\n";

    let third_constant = "down and pass it around,";
    let fourth_constant = "of beer on the wall.\n";

    return `${num} ${start_bottle} ${first_constant} ${num} ${start_bottle} ${second_constant}Take ${end_num > 0 ? "one" : "it"} ${third_constant} ${end_num} ${end_bottle} ${fourth_constant}`;
}

var last_verse = "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
class BeerSong
    def initialize
    end
    def verse(verse_no)
        s1(verse_no) + s2(verse_no)
    end
    def s1(verse_no)
        num_bottles = verse_no > 0 ? verse_no : "No more"
        plural = verse_no == 1 ? "" : "s"
        "#{num_bottles} bottle#{plural} of beer on the wall, " +
        "#{num_bottles.to_s.downcase} bottle#{plural} of beer.\n"
    end
    def s2(verse_no)
        p1 = verse_no > 0 ? "Take #{verse_no == 1 ? "it" : "one"} down and pass it around" : "Go to the store and buy some more"
        num_bottles = verse_no - 1 == 0 ? "no more" : (verse_no - 1 == -1 ? 99 : verse_no - 1)
        plural = verse_no - 1 == 1 ? "" : "s"
        "#{p1}, #{num_bottles} bottle#{plural} of beer on the wall.\n"
    end
    def verses(verse_start, verse_end)
        verse_start.downto(verse_end)
            .map { |index| verse(index) }
            .join("\n")
    end
end
module BookKeeping
    VERSION = 3
end